### ISSUE TITLE
Build failure with `asset_host` config and rack >=2.1.0 with `Content-Length header` error

### ISSUE THREAD

**Issue by lsirivong** [middleman/middleman#2309]
## Expected behavior and actual behavior

`middleman build` succeeds

## Steps to reproduce the problem (from a clean middleman installation)

1. `gem install middleman`
2. `middleman init`
3. add `activate :asset_host, host: 'http://assets.example.com'` to `config.rb`
4. `middleman build --verbose`

Observe build fails.

Additionally, if you downgrade rack you can observe that the build will succeed:

5. `rm -f Gemfile.lock`
6. `echo "gem 'rack', '2.0.8'" >> Gemfile`
7. `middleman build --verbose`

Observe the build succeeds

## Additional information

- Ruby version: ruby 2.6.3p62 (2019-04-16 revision 67580) [x86_64-darwin18]
- Middleman version: 4.3.5
- OS version: Mac OS 10.14.6

Code example:

https://github.com/lsirivong/middleman-example (build fails on master branch)
https://github.com/lsirivong/middleman-example/tree/rack-2.0.8 (build succeeding on this branch)

**Comment by lsirivong** [comment]
Related conversation in rack repo: https://github.com/rack/rack/issues/1472#issuecomment-574428441

**Comment by tomrutgers** [comment]
This occurs without activating `asset_host` as well, when trying to load images.

**Comment by julik** [comment]
Me and my colleague have investigated the issue and the culprit seems to be in the asset hash and asset host rewriting. Basically, the metaprogrammed method called `#rewrite_inline_urls` (which was _very_ hard to find the definition of) does change the contents of resources, but does not change the precomputed `Content-Length` headers for them. We were not able to find how the result of the rewriting feeds back into the Rack response pipeline, so we opted for adding the following to our `config.rb` for Middleman to investigate the issue:

```ruby
class ZapContentLength < Struct.new(:app)
  def call(env)
    s, h, b = app.call(env)
    # The URL rewriters in Middleman do not update Content-Length correctly,
    # which makes Rack-Link flag the responses as having a wrong Content-Length.
    # For building assets this has zero importance because the Content-Length
    # header will be discarded - it is the server that recomputes it. But
    # it does prevent the site from building correctly.
    #
    # The fastest way out of this is to let Rack recompute the Content-Length
    # forcibly, for every response, at retrieval time.
    #
    # See https://github.com/rack/rack/issues/1472
    h.delete('Content-Length')
    [s, h, b]
  end
end

app.use ::Rack::ContentLength
app.use ZapContentLength
```

This forces recomputation of the Content-Length header leading to the correct result. I had serious trouble navigating the Contracts-based inheritance chain for the rewriters, so I hope someone more knowledgeable with Middleman internals can address this.

**Comment by julik** [comment]
I haven't tested pinning Rack to be honest, will check it out when I have time.

That being said, I can see is that the checks for content-length in Rack::Lint are fairly old, the initial implementation close to a decade old, and I haven't seen this failure mode in any of my applications, and the failure only appears when the rewriters in question are applied.

**Comment by speedy1812** [comment]
For me, downgrading to 2.0 wasn't enough. I had to [lock the version to 2.0.8](https://github.com/speedy1812/daysinukraine.com/commit/fdfd9b47c6747812fe05715ff074eb0382d30c82) to resolve the content-length errors I was getting.

**Comment by julik** [comment]
@tdreyno if you can point me in the direction of some architecture document that explains where to find the bits and bobs of the URL rewriting I can put up a patch that fixes the issue properly

**Comment by billofrights** [comment]
Oh my word, @julik, **thank you** for this Content Length fix. I felt like I was going insane.

For further context / for anyone else searching for this bug for CSS (not Rack) reasons: This bug is also tripped by combining `relative_assets` with even one instance of `image-url()` in your styles. This bug prevents the CSS file from being generated.

**Comment by tdreyno** [comment]
@julik The core piece of code is here: https://github.com/middleman/middleman/blob/4.x/middleman-core/lib/middleman-core/core_extensions/inline_url_rewriter.rb

Basically, extensions can register to be able to manipulate the results of an inline url. They are run in  the order that they are added (but only because Ruby's Hash works this way).

Updating the content size is likely something that needs to happen here: https://github.com/middleman/middleman/blob/4.x/middleman-core/lib/middleman-core/core_extensions/inline_url_rewriter.rb#L126

**Comment by julik** [comment]
Thanks @tdreyno but I am not finding this file on master, where did it go?

**Comment by tdreyno** [comment]
Rack support is gone on `master`. The 4.x branch tracks the stable version so patches can be made to it for the 4.x series.

**Comment by daveharrisonnet** [comment]
Is this fixed?

**Comment by julik** [comment]
I've put up a patch in https://github.com/middleman/middleman/pull/2316 - can someone in thread test it? I do not have time to do so at the moment but I believe it should be sufficient.

**Comment by tdreyno** [comment]
If someone who can reproduce the issue can test this PR, I'll merge and get a release out ASAP

**Comment by julik** [comment]
I have reproduced and verified that the PR fixes the issue on our website, so from my side it's good to go @tdreyno

**Comment by tomrutgers** [comment]
I can confirm the PR fixes the problem, I tested it on two different sites.

**Comment by cfillion** [comment]
That PR fixes it here as well.

**Comment by tdreyno** [comment]
Released as `v4.3.6`
